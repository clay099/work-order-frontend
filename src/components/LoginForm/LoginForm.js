import React from "react";
import useFields from "../../hooks/useFields";
import {
	DialogTitle,
	DialogContent,
	TextField,
	DialogActions,
	Button,
} from "@material-ui/core";
import { useHistory } from "react-router-dom";
import { useDispatch } from "react-redux";
import { loginUserWithAPI } from "../../actions/user";
import { loginTradesmenWithAPI } from "../../actions/tradesmen";

/** LoginForm Component
 * @param  {string} userType
 * @param  {function} handleClose
 *
 * Creates formData, handleChange and resetFormData via useFields custom hook
 *
 * Creates a handleSubmit function to submit the login details to API
 *    - If a user is returned from the API save it in Redux State and redirect to the user/tradesmen dashboard
 *    - If an error is returned from the API close Dialog and allow for the feedback SnackBar to provide the user details
 *
 *
 * Renders:
 *    - DialogTitle
 *    - DialogContent TextField input for email
 *    - DialogContent TextField input for password
 *    - Submit button
 */
const LoginForm = ({ userType, handleClose }) => {
	const INITIALSTATE = { email: "", password: "" };
	const { formData, handleChange, resetFormData } = useFields(INITIALSTATE);

	const dispatch = useDispatch();
	const history = useHistory();

	const handleSubmit = async (e) => {
		e.preventDefault();
		let resp;
		if (userType === "User") {
			resp = await dispatch(
				loginUserWithAPI({
					email: formData.email,
					password: formData.password,
				})
			);
		} else {
			resp = await dispatch(
				loginTradesmenWithAPI({
					email: formData.email,
					password: formData.password,
				})
			);
		}
		resetFormData();
		handleClose();
		if (resp.type === "LOGIN_ERROR") {
			// user could not be found. return don't redirect, snackbar should provide user feedback
			return;
		}
		history.push(`/${resp.user_type}`);
	};

	const handleTestSubmit = async (e) => {
		e.preventDefault();
		let resp;
		if (userType === "User") {
			resp = await dispatch(
				loginUserWithAPI({
					email: "user@gmail.com",
					password: "password",
				})
			);
		} else {
			resp = await dispatch(
				loginTradesmenWithAPI({
					email: "tradesmen@gmail.com",
					password: "password",
				})
			);
		}
		resetFormData();
		handleClose();
		if (resp.type === "LOGIN_ERROR") {
			// user could not be found. return don't redirect, snackbar should provide user feedback
			return;
		}
		history.push(`/${resp.user_type}`);
	};

	return (
		<div className="LoginForm">
			<DialogTitle id={`login-form-${userType}`}>
				{userType} Login Form
			</DialogTitle>
			<form>
				<DialogContent>
					<TextField
						required
						autoFocus
						id={`${userType}-email`}
						label="Email Address"
						type="email"
						name="email"
						onChange={handleChange}
						value={formData.email}
					></TextField>
					<TextField
						required
						id={`${userType}-password`}
						label="Password"
						type="password"
						name="password"
						onChange={handleChange}
						value={formData.password}
					></TextField>
				</DialogContent>
				<DialogActions>
					<Button
						type="submit"
						onClick={handleTestSubmit}
						color="secondary"
					>
						Login With Test Account
					</Button>
					<Button
						type="submit"
						onClick={handleSubmit}
						color="primary"
					>
						Login
					</Button>
				</DialogActions>
			</form>
		</div>
	);
};

export default LoginForm;
