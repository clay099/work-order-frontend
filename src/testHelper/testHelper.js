// testHelper.js
import React from "react";
import { render as rtlRender } from "@testing-library/react";
import { createStore } from "redux";
import { Provider } from "react-redux";
// Import your own reducer
import reducer from "../reducers/rootReducer";

function render(
	ui,
	{
		initialState,
		store = createStore(reducer, initialState),
		...renderOptions
	} = {}
) {
	function Wrapper({ children }) {
		return <Provider store={store}>{children}</Provider>;
	}
	return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

const tokenState = { token: "token" };
const userTypeUserState = { userType: "user" };
const userTypeTradesmenState = { userType: "user" };
const idState = { id: 1234 };
const userDetailsState = { userDetails: "userDetails" };
const handleSubmitState = { handleSubmit: jest.fn() };
const handleChangeState = { handleChange: jest.fn() };
const disabledState = { disabled: false };
const disabledEmailState = { disabledEmail: false };
const buttonTextState = { buttonText: "buttonText" };
const formDataState = {
	formData: {
		id: 1,
		firstName: "firstName",
		lastName: "lastName",
		email: "email",
		password: "password",
		newPassword: "newPassword",
		confirmNewPassword: "confirmNewPassword",
	},
};

// re-export everything
export * from "@testing-library/react";
// override render method
export {
	render,
	tokenState,
	userTypeUserState,
	userTypeTradesmenState,
	idState,
	userDetailsState,
	handleSubmitState,
	handleChangeState,
	disabledState,
	disabledEmailState,
	buttonTextState,
	formDataState,
};
