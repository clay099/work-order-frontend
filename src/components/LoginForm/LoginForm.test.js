import React from "react";
import LoginForm from "./LoginForm";
import { render } from "../../testHelper/testHelper";
console.error = jest.fn();

const params = {
	userType: "user",
	handleClose: jest.fn(),
};

describe("LoginForm Component", () => {
	// smoke test
	test("renders without crashing", () => {
		render(<LoginForm {...params} />);
	});

	// snapshot test
	test("matches snapshot", () => {
		const { asFragment } = render(<LoginForm {...params} />);
		expect(asFragment).toMatchSnapshot();
	});

	test("value to be in the document", () => {
		const { getByText } = render(<LoginForm {...params} />);
		const value = getByText("Login");
		expect(value).toBeInTheDocument();
	});
});
