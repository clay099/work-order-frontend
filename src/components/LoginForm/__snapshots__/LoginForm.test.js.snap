// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm Component matches snapshot 1`] = `[Function]`;
